[PLATFORM-422] Module – styling pass

mondoreale opened the pull request:
I did more than I planned. It looks good tho!

![1_tabbing_through_port_controls](https://user-images.githubusercontent.com/320066/56213580-98e65380-605c-11e9-96c5-f1996892ba00.gif)
_(^^ Tabbing through inputs)_

![editing_port_values](https://user-images.githubusercontent.com/320066/56213583-9ab01700-605c-11e9-8589-2427941eb8fb.gif)
_(^^ Filling out inputs)_

@timoxley @tumppi please throw an eye. It's quite hefty, but it is what it is. Already looking like something we could take screenshots of.

I adjusted `ColorPicker` component to fit the new layout, changed it just slightly. I'm gonna have to do the same to `StreamSelector` cause it's a bit off atm.

Tomorrow I'm gonna go through @mattatgit's list again and make sure everything is done.

Outstanding tasks:
- [x] Adjust `StreamSelector` to match other controls and the the design

mattatgit commented:
@mondoreale That *is* a beast. But looks really good, thank you :)

timoxley commented on app/src/editor/canvas/components/Ports/Port/index.jsx:
```diff
@@ -0,0 +1,174 @@
+    }, [])
+
+    const onDocumentClick = useCallback((e: SyntheticMouseEvent<EventTarget>) => {
+        if (contextMenuTarget && e.target instanceof HTMLElement) {
```
Is it possible that `e.target` is an `SVGElement` rather than an `HTMLElement`? `classList` comes from `Element` which is the base class for both.

timoxley replied:
Also, is it possible that `e.target` comes from a different realm i.e. different `HTMLElement` instance? e.g. an iframe.

mondoreale replied:
We may have a module that has an `iframe`. Added 8b4611f and changed `HTMLElement` to `Element`.

timoxley commented on app/src/editor/canvas/components/Ports/Port/index.jsx:
```diff
@@ -0,0 +1,174 @@
+    }, [port.id, onValueChangeProp])
+
+    useEffect(() => {
+        window.addEventListener('mousedown', onDocumentClick)
```
`onWindow` click perhaps, since it's not actually `document.addEventListener`?

timoxley commented on app/src/editor/canvas/components/Ports/Port/index.jsx:
```diff
@@ -0,0 +1,174 @@
+
+        return () => {
+            window.addEventListener('mousedown', onDocumentClick)
+            window.addEventListener('keydown', onKeyDown)
```
Should this be `removeEventListener`?

timoxley commented on app/src/editor/canvas/components/Ports/Port/index.jsx:
```diff
@@ -0,0 +1,174 @@
+
+    return (
+        <DragDropContext.Consumer>
+            {({ isDragging, data }) => {
```
`const { isDragging, data } = useContext(DragDropContext)`?

timoxley commented on app/src/editor/canvas/components/Ports/Value/Map/index.jsx:
```diff
@@ -0,0 +1,75 @@
+            .filter(([key = '']) => key.trim())
+            // Turn it into an Object.
+            .reduce((memo, [key, val = '']) => {
+                const v: any = val
```
blegh

timoxley commented on app/src/editor/canvas/components/Ports/Value/Map/index.jsx:
```diff
@@ -0,0 +1,75 @@
+}
+
+const Map = ({ disabled, onChange, value }: Props) => {
+    const [entries, setEntries] = useState(Object.entries(value))
```
It appears the `entries` state is never updated if the `value` prop changes?

timoxley replied:
I have a feeling this may be the cause of undo/redo not working.

timoxley commented on app/src/editor/canvas/components/Ports/Value/Select/index.jsx:
```diff
@@ -0,0 +1,56 @@
+                    ))}
+                </select>
+                {/* `select` holding a currently selected value. This hidden (`visibility: hidden`) control
+                    dictates the width of the actual (visible) control above. */}
```
cool

timoxley commented on app/src/editor/canvas/components/Ports/Value/Text/index.jsx:
```diff
@@ -0,0 +1,33 @@
+    return (
+        <EditableText
+            {...props}
+            className={cx(styles.root, {
```
should allow a custom `className` to be supplied

timoxley replied:
unless the intention is if supplying a custom `className` then the default classes will not apply at all?

timoxley commented on app/src/editor/canvas/components/Ports/Value/index.jsx:
```diff
@@ -0,0 +1,89 @@
+        default:
+    }
+    return 'text'
+}
```
cool

timoxley commented on app/src/editor/canvas/components/Ports/index.jsx:
```diff
@@ -0,0 +1,63 @@
+                        api={api}
+                        canvas={canvas}
+                        /* eslint-disable react/no-array-index-key */
+                        key={port.id + index}
```
this shouldn't need the index unless you're intentionally remounting when ports are reordered?

timoxley replied:
have you seen non-unique port ids?

mondoreale replied:
Copied that piece over. Wasn't sure about the uniqueness of them ids. Will try w/o `index`.

timoxley replied:
> Copied that piece over. 

@mondoreale ah, I think this may have been a basic protection against placeholder keys colliding with a real port's key. No longer required.

timoxley commented on app/src/editor/canvas/components/Ports/index.jsx:
```diff
@@ -0,0 +1,63 @@
+                        api={api}
+                        canvas={canvas}
+                        /* eslint-disable react/no-array-index-key */
+                        key={port.id + index}
```
same here

timoxley commented on app/src/shared/assets/stylesheets/variables.css:
```diff
@@ -1,5 +1,8 @@
   --fontSize: 1rem;
+  --mono: 'IBM Plex Mono', monospace;
+  --sans: 'IBM Plex Sans', sans-serif;
+  --um: 16px; /* Wishful thinking: Future 1rem? — Mariusz */
```
what is `um`?

mondoreale replied:
Unit of measurement. `#namingishard`

timoxley commented on app/src/shared/components/EditableText/index.jsx:
```diff
@@ -20,37 +22,48 @@ const EditableText = ({
+                // In order to allow shift-tabbing through interactive elements
+                // we can't let the span be focusable when the input is.
+                tabIndex: (hasFocus ? -1 : 0),
+            } : {})}
```
cool

timoxley commented on app/src/shared/components/TextControl/index.jsx:
```diff
@@ -45,7 +45,7 @@ const TextControl = ({
     const ref = innerRef || el
     const reverted: Ref<boolean> = useRef(false)
-    const [value, setValue] = useState(valueProp || '')
+    const [value, setValue] = useState(valueProp == null ? '' : valueProp)
```
There's a few other places where `setValue(valueProp)` is called, should it apply the same default logic in those places too?

mondoreale replied:
I'm pretty sure in other places `setValue` is called always with a string. Will double-check.

timoxley commented on app/webpack.config.js:
```diff
@@ -217,6 +217,8 @@ module.exports = {
                     // skip MD documentation
                     'src/docs/docsEditingGuide.md',
+                    // skip sketch files
+                    '**/*.sketch',
```
sketch files?

mondoreale replied:
Commit message for 61404a8d234f66a8c78d6c9aff297a7a4987a27e:
> **Let .sketch files be unused**
> I tend to keep a local sketch file for components I work with (slices, simplified svgs, etc). Don’t want them to be catched by UnusedFilesWebpackPlugin.

😃

timoxley replied:
👍

timoxley commented:
According to this mock, map params should be aligned on some kind of grid:

![image](https://user-images.githubusercontent.com/43438/56268412-dc57c500-6123-11e9-8389-a1aedb74341d.png)

Currently:

![image](https://user-images.githubusercontent.com/43438/56268466-fdb8b100-6123-11e9-8f89-bfc3b773780c.png)

mondoreale commented on app/src/shared/components/TextControl/index.jsx:
```diff
@@ -45,7 +45,8 @@ const TextControl = ({
     const reverted: Ref<boolean> = useRef(false)
-    const [value, setValue] = useState(valueProp || '')
+    const sanitizedValue = valueProp == null ? '' : valueProp
+    const [value, setValue] = useState(sanitizedValue)
```
@timoxley, you were right. Same logic does apply to other places (all `setValue(valueProp)`). Good catch!

mondoreale commented:
@timoxley I fixed
- `Esc` issue (EditableText) in ed17127, and
- Map's column alignment in b5f9111.

Btw, the CI builds behave consistently! 👍

mondoreale commented:
Spotted a vertical misalignment, hold on.

mondoreale commented:
Fixed. Taking care of StreamSelector now.

mondoreale commented:
Btw, @timoxley @mattatgit, Map's cells:
<img width="480" alt="Screenshot 2019-04-17 12 57 26" src="https://user-images.githubusercontent.com/320066/56283649-a78f4200-6112-11e9-8b5c-5c54ee4ef020.png">

Columns are resized dynamically based on content. In the image above `Params` and `Headers` are separate Maps and work independently.

mondoreale commented:
Alright guys. Looks like the time has come.

@timoxley please throw an eye on it again real quick.

@tumppi I modified `StreamSelector` and `ColorPicker` components for this. Not much, but not little either. ;) Please scroll through it too. The more  👀 the better.

Hope you guys like it.

timoxley commented on app/src/editor/canvas/components/Ports/Port/index.jsx:
```diff
@@ -0,0 +1,178 @@
+                    name="noRepeat"
+                    onToggle={onOptionToggle}
+                />
+            )}
```
Would be good to make sure all of these `conditional && <Component />` were forcibly coerced to booleans i.e. `!!conditional && <Component />` so we can't possibly accidentally render a `0`, ever. Figuring out exactly which code path happened to generate a zero can be very time consuming. I know `flow` should prevent this, but `flow` isn't perfect and the cost to prevent this from ever happening is just adding `!!` so it seems to me a low cost/high yield habit.

timoxley commented on app/src/editor/canvas/components/Ports/Value/Color/index.jsx:
```diff
@@ -0,0 +1,33 @@
+import ColorPicker from '$editor/shared/components/ColorPicker'
+import { type CommonProps } from '..'
+
+type Props = CommonProps & {}
```
` & {}`?

timoxley commented on app/src/editor/canvas/components/Ports/index.jsx:
```diff
@@ -0,0 +1,61 @@
+import cx from 'classnames'
+
+import Port from './Port'
+import styles from './ports.pcss'
```
>  error  Casing of ./ports.pcss does not match the underlying filesystem

timoxley commented on app/src/shared/components/EditableText/editableText.pcss:
```diff
@@ -30,13 +30,25 @@
 
+.idle {
+  cursor: text;
+}
```
I note that this cursor looks weird when dragging the module around by the header. Maybe the editable text's width in the header should be `min-content` or perhaps the cursor should get changed if a drag is in progress.

mondoreale replied:
https://streamr.atlassian.net/browse/PLATFORM-533

mondoreale commented:
@timoxley 
- 202d34a – fixed undo/redo for `Map` and `Color`, and
- f7ed56b – auto-focus `name` of newly added `Map` entries (no refs!)

I'm _hooked_!

timoxley commented:
Also note the last row with any content has no `-` button.  The mocks aren't entirely clear how this should work, but the way I interpreted it:

* Empty key/value inputs are always visible as the last row. This row is automatically created after the last row with content. The `+` button doesn't really do anything (just focuses the key input) and could even be removed.
* Always displaying an empty row at the end allows every row with content to have a `-` button.
* Also means user doesn't need to keep activating the + button to get new rows, they can just keep hitting tab and filling in values.

timoxley commented:
Also should probably have a max-width on the keys & values.

![image](https://user-images.githubusercontent.com/43438/56400664-a3386580-6287-11e9-90ae-b33667e3d8a9.png)

mondoreale commented:
@timoxley, further Map param issues are connected strictly to the undo/redo. We want to store the intermediate map param value for undoing and redoing, but it means we have to store the actual sanitized value – map param's value is an object, i.e. it won't allow dupped keys and the order of keys can change.

Ideally map param's value is an array of objects or arrays, e.g. `Array<{name: string, value: mixed}>` and HTTP module or any module utilizing map params sorts the final value out on its own.

mondoreale commented:
Said it too fast. There may be a way to make it less confusing with what we have… Lemme try.

timoxley commented:
@mondoreale yeah I believe the way I tried to make it work previously is by manipulating the value as an Object.entries() array while editing then coercing to/from an object as data enters/leaves the component, which isn't so far from what you have already.

mondoreale commented:
Like fixing an engine, sir. _Try it now!_ 🔧 

- I listened and removed that "+" icon. It really feels redundant.
- Fixed redo/undo. Duplications behave as 1 entry. You're not gonna see reordering unless you reload (no way around undetermined order of Object's props).
- Added a limit to `Text` component width

It should fly nicely now. 🚀

timoxley commented:
@mondoreale Good enough. Still some definite weirdness with duplicate/empty keys & undo/redo, but I think this is overall basically workable, let's improve the map param handling in a new PR.

timoxley commented on app/src/editor/canvas/components/Ports/Value/Map/index.jsx:
```diff
@@ -2,52 +2,81 @@
                 return {
                     ...memo,
-                    [key.trim()]: v.trim(),
+                    [k]: values[k].trim() || '',
```
How can `values[k].trim()` be falsey unless it's already `''`?

mondoreale commented:
Yep, dup/empty key weirdness included… ;) Let's address as a bug or something, once someone reports it. Leaving it be for now.

mattatgit commented:
@mondoreale Just to note, the subtle animation on module selection & deselection is 👌👍
